Add e2e test for GRPC autoscaling

shashwathi opened the pull request:
Add test thats asserts whether GRPC app autoscales(1->n) with traffic, scales
down to 0 and then from 0-n instances. Follows similar pattern to existing autoscale test

<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Fixes #6356

shashwathi commented:
/test pull-knative-serving-smoke-tests

shashwathi commented:
Both of your options makes sense to me.

> Alternatively, maybe we can reduce the scope of the gRPC autoscaling tests to just "does gRPC autoscaling work", and let autoscale_test.go handle the more rigorous testing?

So if I understand that comment correctly, the reduced scope for e2e test would be that app scales on GRPC traffic. 

any other thoughts ? 
@markusthoemmes @mattmoor @vagababov

tcnghia commented:
/cc @ahmetb

tcnghia commented:
Is there a way we can verify that the responses reach more than one pod?  There was a report that we did autoscaling, but all the requests are going to the same pod.

I guess we could modify the test image to return a unique value in their signature.  Or if there are response headers that tell us the identity of the server we can use that.

nimakaviani commented:
@tcnghia I had to implement a similar concept for testing with websockets where the response includes `os.Hostname()` in its body. The PR is still under review but throwing it here in case it helps: https://github.com/knative/serving/pull/6606

tcnghia commented:
@nimakaviani that is awesome.  Thanks for the pointer.

shashwathi commented:
@tcnghia : I updated e2e test to include hostname in response like @nimakaviani suggested and I noticed that all responses are generated from single host. 
I have pushed that change in this PR and e2e test should fail for the same reason. Please validate if my testing method is correct and if that is correct then let me know how do you want to handle this.

@tanzeeb : I have refactored the code to reuse autoscale function. Let me know what you think.

shashwathi commented:
/assign @markusthoemmes

shashwathi commented:
I am wondering if it makes more sense to track this(load balance GRPC traffic during autoscale) bug/feature in separate issue. The scope of this PR was to test whether app autoscales with GRPC traffic and I dont want to scope creep on this PR.

tcnghia commented:
@shashwathi Yes, we can track this bug in a separate issue.  Please comment out the assertion that would expose the bug, and add some comments to the issue so that the person working on the issue can just uncomment to reproduce the issue.

Thanks a lot!

tcnghia commented:
/lgtm

shashwathi commented:
/assign @vagababov

vagababov commented on test/e2e/grpc_test.go:
```diff
@@ -92,13 +96,108 @@ func unaryTest(t *testing.T, resources *v1a1test.ResourceObjects, clients *test.
+		if err != nil {
+			return err
+		}
+		knownHosts := make(map[interface{}]interface{})
```
why do we need a second map? `uniqueHost` is a map.

shashwathi replied:
Cannot count the length of sync map.

vagababov replied:
You can, just `i++` in the range call.

vagababov commented on test/e2e/grpc_test.go:
```diff
@@ -92,13 +96,108 @@ func unaryTest(t *testing.T, resources *v1a1test.ResourceObjects, clients *test.
+	var grp errgroup.Group
+
+	grp.Go(func() error {
+		uniqueHost, err := generateGRPCTraffic(ctx.t, 50, host, domain)
```
```suggestion
		uniqueHosts, err := generateGRPCTraffic(ctx.t, 50, host, domain)
```

vagababov commented on test/e2e/grpc_test.go:
```diff
@@ -92,13 +96,108 @@ func unaryTest(t *testing.T, resources *v1a1test.ResourceObjects, clients *test.
+			knownHosts[k] = v
+			return true
+		})
+		// This assertion checks whether responses are coming from multiple hosts
```
rm commented code?

shashwathi replied:
refer comment : https://github.com/knative/serving/pull/6618#issuecomment-579909698

vagababov replied:
Yeah, remove the code and then re-introduce it when the time comes.

markusthoemmes replied:
Well if we remove that code, we can remove all the uniqueHost map stuff with it. We only do that to facilitate this check.

shashwathi replied:
yes

vagababov commented on test/test_images/grpc-ping/main.go:
```diff
@@ -31,12 +32,20 @@ func pong(req *ping.Request) *ping.Response {
 	log.Printf("Received ping: %v", req.Msg)
-
-	resp := pong(req)
+	time.Sleep(1 * time.Second)
```
what is this sleep for?

shashwathi replied:
Without delay in response GRPC app does not scale.

vagababov replied:
Make this an option :) e.g.extend gRPC API to haven an `optional sleepSecs int = 2;`

vagababov replied:
Or env var for the image.

shashwathi replied:
Okay I can update the app to make it optional

tanzeeb commented:
@googlebot I consent.

vagababov commented on test/e2e/autoscale_test.go:
```diff
@@ -277,6 +277,54 @@ func numberOfPods(ctx *testContext) (float64, error) {
+	for {
+		select {
+		case <-timer:
+			// Each 2 second, check that the number of pods is at least `minPods`. `minPods` is increasing
```
```suggestion
			// Every 2 seconds, check that the number of pods is at least `minPods`. `minPods` is increasing
```

vagababov replied:
Second sentence doesn't compile for me gramatically, I'd reformulate it.

shashwathi replied:
I moved that sentence near the relevant code section

vagababov commented on test/e2e/grpc_test.go:
```diff
@@ -37,7 +41,11 @@ import (
-type grpcTest func(*testing.T, *v1a1test.ResourceObjects, *test.Clients, string, string)
+const (
+	grpcContainerConcurrency = 1.0
+)
```
`const grpcContainerConcurrency = 1.0`

vagababov commented on test/e2e/grpc_test.go:
```diff
@@ -74,31 +82,106 @@ func dial(host, domain string) (*grpc.ClientConn, error) {
+func unaryTest(t *testing.T, resources *v1a1test.ResourceObjects, clients *test.Clients, names test.ResourceNames, host, domain string) {
 	t.Helper()
 	t.Logf("Connecting to grpc-ping using host %q and authority %q", host, domain)
+	msg := "Hello!"
```
```suggestion
	const msg = "Hello!"
```

vagababov commented on test/e2e/grpc_test.go:
```diff
@@ -74,31 +82,106 @@ func dial(host, domain string) (*grpc.ClientConn, error) {
 	t.Logf("Connecting to grpc-ping using host %q and authority %q", host, domain)
+	msg := "Hello!"
+	if err := pingGRPC(host, domain, msg); err != nil {
+		t.Fatalf("Error %#v", err)
```
```suggestion
		t.Errorf("gRPC ping = %v", err)
```

vagababov commented on test/e2e/grpc_test.go:
```diff
@@ -74,31 +82,106 @@ func dial(host, domain string) (*grpc.ClientConn, error) {
+		})
+	}
+	if err := grp.Wait(); err != nil {
+		return fmt.Errorf("Error processing requests %v", err)
```
```suggestion
		return fmt.Errorf("error processing requests %v", err)
```

vagababov commented on test/e2e/grpc_test.go:
```diff
@@ -74,31 +82,106 @@ func dial(host, domain string) (*grpc.ClientConn, error) {
 
 	if got.Msg != want.Msg {
-		t.Errorf("Response = %q, want = %q", got.Msg, want.Msg)
+		return fmt.Errorf("Response = %q, want = %q", got.Msg, want.Msg)
```
```suggestion
		return fmt.Errorf("response = %q, want = %q", got.Msg, want.Msg)
```

shashwathi commented:
@vagababov : Addressed your comments. Take another look

shashwathi commented:
@markusthoemmes @vagababov @tcnghia @nak3 : Ready for another review

tcnghia commented:
/lgtm
/approve
